### ISSUE TITLE
Allow ResourceShapes to extend other ResourceShapes

### ISSUE THREAD

**Issue by jamsden** [jamsden/oslc-core-test#65]
Many of the existing OSLC domain vocabularies are fairly simple, defining a relatively small number of RDFS classes, and rarely (if ever) use subClassOf. As a result an instance of a domain class can have an oslc:instanceShape that references a simple ResourceShape.

However, some vocabularies, such as those derived from UML, SysML, PLE, etc. may have much richer vocabularies that have deep inheritance hierarchies. Although it is possible to create A ResourceShape for each subclass, this would result in a very large number of repeated properties as the properties inherited from superclasses would need to be copied down into the ResourceShape for each subclass in order to fully specify the resource shape.

I is therefore desirable to have some means of "inheriting" constraints from some other ResourceShape(s), or extend a ResourceShape with constraints defined in some other shape - a chain of constraints like prototypes in JavaScript.

**Comment by ndjc** [comment]
Allowing some way to extend a resource shape provides another benefit: it can have the effect of injecting a property unknown to the author of the original shape. For example, suppose you have a requirement, conforming to the OSLC RM specification and its resource shapes. The shape for this resource comes from the tool publishing that resource; let us name that RM\\\_App. Now, another application XX\\\_App decides to publish a triple whose subject is that requirement, but the predicate and object are something new, defined by XX\\\_App. We presume XX\\\_App cannot modify the shape produced by RM\\\_App. How can that triple get known? If the object of this triple was a resource owned by XX\\\_App then one approach would be to extend shapes to allow reverse property definitions, and to add a reverse property for the triple being added to the requirement. However, this would not work if the object of the triple was a literal. Another, more general, approach is to allow XX\\\_App to publish a shape that extends the shape defined by RM\\\_App. If necessary, XX\\\_App could also add a new rdf:type triple to the requirement from RM\\\_App, so the requirement is covered by the oslc:describes property of the shape from XX\\\_App.

**Comment by ndjc** [comment]
I agree that if the new shape describes only part of the resource - just the extension properties and not the base properties - then using oslc:describes and a new type would not be appropriate. In that case, it would be better to add a new predicate (oslc:extends?).
